### ISSUE TITLE
SPIKE: Investigate automatic language detection for literals

### ISSUE THREAD

**Issue by jbaxmeyer** [LD4P/sinopia_editor#2876]
I typed in two strings of text in Field 1, one in Chinese script and the other in its romanized form. The default language indication is English. I would then change it to Chinese for both.
In comparison Google Translate https://translate.google.com/ can detect language/script automatically, including the romanized form. 
Can we automate language detection here likewise? Yes, it always needs human proofreading but when the machine gets it right keystrokes and time are saved.
<img width="483" alt="autolangdetect" src="https://user-images.githubusercontent.com/7747876/63172236-39e51b00-c00b-11e9-897c-e2c66540e434.PNG">

**Comment by NancyL** [comment]
Can one automate language for transliterated data such as in the first part of Jennifer's example above as well as the original script?

**Comment by michelleif** [comment]
as part of this spike, could we investigate also whether script can be automatically detected, along with language?

**Comment by jcoyne** [comment]
Isn't the script written into the characters themselves? That seems like it ought to be a separate ticket as AWS Comprehend is not about script.  

When we detect multiple languages what should the behavior be?

**Comment by justinlittman** [comment]
Do you know what "Reliable" means?

**Comment by jcoyne** [comment]
@justinlittman ignore that column, this demonstration comes from a Azure service. The point is that we're given a list of languages with a "score" that indicates how likely this answer is to be correct.

**Comment by justinlittman** [comment]
Do you know what codes are used for the languages?

**Comment by jcoyne** [comment]
@justinlittman It's just English strings. They don't use codes.

**Comment by jcoyne** [comment]
> Amazon Comprehend identifies the language using identifiers from RFC 5646 — if there is a 2-letter ISO 639-1 identifier, with a regional subtag if necessary, it uses that. Otherwise, it uses the ISO 639-2 3-letter code.

https://docs.aws.amazon.com/comprehend/latest/dg/how-languages.html

**Comment by jcoyne** [comment]
To proceed farther with this, we should have a discussion about our approach/architecture for passing credentials to a client or proxying this service.

**Comment by justinlittman** [comment]
The proposed approach is to proxy behind the Sinopia API. (Other AWS services that require credentials are currently abstracted by the API and are not called directly from Sinopia Editor.)

**Comment by justinlittman** [comment]
Is there some sort of a testbed that we create to allow catalogers to evaluate the language detection?

**Comment by jcoyne** [comment]
So, this might be bad news for the transliterated Chinese given in the example:
```
16:17 $ AWS_PROFILE=dlss-devRole aws comprehend detect-dominant-language --text "zhe yang zhen hao"
{
    "Languages": [
        {
            "LanguageCode": "jv",
            "Score": 0.18218006193637848
        },
        {
            "LanguageCode": "id",
            "Score": 0.27455785870552063
        },
        {
            "LanguageCode": "en",
            "Score": 0.3142596185207367
        }
    ]
}
```

**Comment by jcoyne** [comment]
@justinlittman sure, what do you want the test bed to look like?

**Comment by justinlittman** [comment]
Open to suggestions, but could be a simple API gateway app that invokes a lambda and uses a form served from S3. I don't know if this is helpful, but see the files removed in https://github.com/LD4P/rdf2marc/commit/7243a525bd43179bde67e3b562fa313e1c712e9e for how similar was done with rdf2marc.

**Comment by jcoyne** [comment]
See https://github.com/LD4P/language-detection, currently waiting on Ops to grant `comprehend:DetectDominantLanguage` to the `LanguageDetectionDemoExecutionRole`

**Comment by michelleif** [comment]
to those following this ticket, a test bed is here https://ld4p.github.io/language-detection/ to enter a string and get a detected language plus a  “likelihood of correctness” score between 0 and 1, where 1 is the most likely. try it out to see if this would be a welcome feature to add to Sinopia, to automatically tag language of entered literals

**Comment by jcoyne** [comment]
@michelleif I think the bigger question to answer using the test bed is, what do we do when the results are somewhat ambiguous?  What score is required before we think of it as good enough?

**Comment by michelleif** [comment]
thanks @jcoyne i'll ask then when i follow up with people

**Comment by justinlittman** [comment]
I think the feature would be "suggest" the tag, rather than "automatically" tag. Or, possibly, it could suggest multiple tags. But I think the goal is to make it quicker for a cataloger to select the correct tag (compared to having to open the language modal, type something in, make a choice, close the modal, etc.)

**Comment by justinlittman** [comment]
Declaring victory on the investigation. Will ticket separately if decide to proceed.

**Comment by ndushay** [comment]
er ... I could have told you that the strings being checked are too short, in general, for language detection.  Also, a number of scripts are used in multiple languages, creating the ambiguity.   The hebrew script is used for Hebrew and Yiddish.  The arabic script is used for a bunch of languages.   The Chinese script is used for at least 3 languages.  Longer text gives a higher likelihood of disambiguating.   Most text going into a single cataloging value is ... too short.
